Federated Model (seq-to-seq)/train_model.py

```diff
@@ -249,53 +323,78 @@ def create_keras_model(arg,
 
 
 def preprocess(dataset, arg):
-    """Shuffles and batches the dataset.
-    """
-    return (dataset.shuffle(BUFFER_SIZE).batch(arg.batch_size,
-                                               drop_remainder=False))
+    """Preprocessing function fot the training data
 
+    Args:
+    dataset: The tensor containing the training data.
+    arg: The output of the parser.
+    epochs_per_rounds: Number of epochs per rounds of federated training.
 
-def make_federated_data(client_data, client_ids, arg):
-    """Genrates the required federated input.
-    """    
-    return [
-        preprocess(client_data.create_tf_dataset_for_client(str(x)), arg)
-        for x in client_ids
-    ]
+    Returns:
+    The pre-processed dataset.
+    """
+    return (dataset.repeat(arg.epochs_per_round).shuffle(BUFFER_SIZE).batch(
+        arg.batch_size, drop_remainder=False))
 
 
-def masked_slot_loss(y_true, y_pred):
-    """Defines the loss function to be used while training the model.
-    The loss is defined as the sum of the intent loss and the slot loss
-    per token. Masking is used to not compute the loss on the padding tokens.
+def get_optimizers(arg):
+    """Returns the optimizers for the server and the clients based
+    on the input arguments.
 
     Args:
-    slot_real: The ground truth for the slots.
-    intent_real: The ground through for the intents.
-    slot_pred: The slots predicted by the model.
-    intent_pred: The intents predicted by the model.
-    intent_loss_objective: The objective used to compute the intent loss.
-    slot_loss_objective: The objective used to compute the slot loss.
+    arg: The output of the parser.
 
     Returns:
-    The total loss.
+    The server and client optimizer.
     """
-    loss_objective = tf.keras.losses.SparseCategoricalCrossentropy(
-        from_logits=True, reduction='none')
-    mask = tf.math.logical_not(tf.math.equal(y_true, 0))
-    slot_losses = loss_objective(y_true, y_pred)
+    server_opt_cls = _SUPPORTED_OPTIMIZERS.get(arg.server_optimizer)
+    client_opt_cls = _SUPPORTED_OPTIMIZERS.get(arg.client_optimizer)
+
+    if arg.server_optimizer == 'sgd':
+        server_opt = lambda: server_opt_cls(learning_rate=arg.server_lr,
+                                            momentum=arg.momentum)
+    else:
```

Comment:
better to be more explicit, i.e elif arg.server_optimizer in ['adam', 'adagrad']:
And add a not support error in the else clause.